impeller/renderer/backend/vulkan/command_pool_vk.cc

```diff
@@ -3,166 +3,190 @@
 // found in the LICENSE file.
 
 #include "impeller/renderer/backend/vulkan/command_pool_vk.h"
+#include <memory>
+#include <optional>
+#include <utility>
+#include "fml/macros.h"
+#include "fml/thread_local.h"
+#include "fml/trace_event.h"
+#include "impeller/renderer/backend/vulkan/resource_manager_vk.h"
+#include "vulkan/vulkan_structs.hpp"
 
-#include <map>
-#include <unordered_map>
-#include <vector>
+namespace impeller {
 
-#include "flutter/fml/thread_local.h"
-#include "impeller/base/thread.h"
-#include "impeller/renderer/backend/vulkan/context_vk.h"
+// Holds the command pool in a background thread, recyling it when not in use.
+class BackgroundCommandPoolVK {
+ public:
+  BackgroundCommandPoolVK(BackgroundCommandPoolVK&&) = default;
+
+  explicit BackgroundCommandPoolVK(
+      vk::UniqueCommandPool&& pool,
+      std::vector<vk::UniqueCommandBuffer>&& buffers,
+      std::weak_ptr<CommandPoolRecyclerVK> recycler)
+      : pool_(std::move(pool)),
+        buffers_(std::move(buffers)),
+        recycler_(std::move(recycler)) {}
+
+  ~BackgroundCommandPoolVK() {
+    auto const recycler = recycler_.lock();
+    if (!recycler) {
+      return;
+    }
+    recycler->Reclaim(std::move(pool_));
+  }
 
-namespace impeller {
+ private:
+  FML_DISALLOW_COPY_AND_ASSIGN(BackgroundCommandPoolVK);
 
-using CommandPoolMap = std::map<uint64_t, std::shared_ptr<CommandPoolVK>>;
-FML_THREAD_LOCAL fml::ThreadLocalUniquePtr<CommandPoolMap> tls_command_pool;
+  vk::UniqueCommandPool pool_;
 
-static Mutex g_all_pools_mutex;
-static std::unordered_map<const ContextVK*,
-                          std::vector<std::weak_ptr<CommandPoolVK>>>
-    g_all_pools IPLR_GUARDED_BY(g_all_pools_mutex);
+  // These are retained otherwise it is a Vulkan thread-safety violation
+  // because the buffers and the originating pool do not belong to the same
+  // thread.
```

Comment:
nit: these are retained because the destructor of the C++ UniqueCommandBuffer wrapper type will attempt to reset the cmd buffer, and doing so may be a thread safety violation as this may happen on the fence waiter thread.

Reply:
Thanks. Changed!